### PR TITLE
Feature/extrinsic deserialization runtime api shuffling

### PR DESCRIPTION
**Pull request by mateuszaaa**
Extrinsic deserialization using dedicated runtime api & generic extrinsic shuffler used on both node and client side 

That PR is done against `feature/extrinsic_deserialization_runtime_api` in order to separate forked code from actual development - that should make review a bit easier:
- forking sc-basic-authorship == 0.8.0
- forking sc-transaction-pool == 2.0.0

### CODE REVIEW

File: client/basic-authorship/Cargo.toml [hunk 1]
```diff
@@ -17,21 +17,23 @@ codec = { package = "parity-scale-codec", version = "1.3.4" }
...
+prometheus-endpoint = { package = "substrate-prometheus-endpoint", version = "0.8.0"}
+sp-api = { version = "2.0.0"}
+sp-runtime = { version = "2.0.0"}
+sp-core = { version="2.0.0"}
+sp-blockchain = { version = "2.0.0"}
+sc-client-api = { version = "2.0.0"}
+sp-consensus = { version = "0.8.0"}
+sp-inherents = { version = "2.0.0"}
+sc-telemetry = { version = "2.0.0"}
+sp-transaction-pool = { version = "2.0.0"}
```
Comment (gleb-urvanov) [inline client/basic-authorship/Cargo.toml:29]: Why altered to use crates form repo version while we still have local ones?
If none of them are modified by us, we should remove them.

File: client/basic-authorship/Cargo.toml [hunk 2]
```diff
@@ -17,21 +17,23 @@ codec = { package = "parity-scale-codec", version = "1.3.4" }
...
+extrinsic-info-runtime-api = { version = "2.0.0", path = "../../primitives/extrinsic-info-runtime-api" }
+sc-proposer-metrics = { version = "0.8.0"}
 tokio-executor = { version = "0.2.0-alpha.6", features = ["blocking"] }
+sc-transaction-pool = { path='../transaction-pool', version = "2.0.0"}
 
 [dev-dependencies]
-sc-transaction-pool = { version = "2.0.0", path = "../../client/transaction-pool" }
-substrate-test-runtime-client = { version = "2.0.0", path = "../../test-utils/runtime/client" }
+#sc-transaction-pool = { path='../transaction-pool', version = "2.0.0"}
+# substrate-test-runtime-client = { version = "2.0.0"}
```
Comment (gleb-urvanov) [inline client/basic-authorship/Cargo.toml:38]: remove dead dependencies

File: client/service/src/client/client.rs [hunk 3]
```diff
@@ -289,6 +289,7 @@ where
 	E: CallExecutor<Block>,
 	Block: BlockT,
 	Block::Header: Clone,
+	// <Self as Client<B, E, Block, RA>>::Api: ExtrinsicInfoRuntimeApi<Block> ,
```
Comment (gleb-urvanov) [inline client/service/src/client/client.rs:291]: Remove dead code

File: client/block-builder/src/lib.rs [hunk 4]
```diff
@@ -58,298 +58,308 @@ use sc_client_api::backend;
...
+    pub fn build(
+        mut self,
+    ) -> Result<BuiltBlock<Block, backend::StateBackendFor<B, Block>>, ApiErrorFor<A, Block>> {
+        let block_id = &self.block_id;
+
+        let extrinsics = self.extrinsics.clone();
+        let parent_hash = self.parent_hash;
+        let extrinsics_hash = BlakeTwo256::hash(&extrinsics.encode());
+
+        //FIXME
```
Comment (gleb-urvanov) [inline client/block-builder/src/lib.rs:218]: From my point of view, the following code is fixed now, so //fixme comment is no longer necessary to keep

File: client/service/src/client/client.rs [hunk 5]
```diff
@@ -929,13 +930,19 @@ where
...
-							let mut shuffled_extrinsics = previous_block_extrinsics.clone();
-							info!("seed: {:?}", extrinsics_hash.to_fixed_bytes());
-							let mut rng: StdRng =
-								SeedableRng::from_seed(extrinsics_hash.to_fixed_bytes());
-							shuffled_extrinsics.shuffle(&mut rng);
+                            let seed: StdRng = SeedableRng::from_seed(extrinsics_hash.to_fixed_bytes());
+                            let shuffled_extrinsics = extrinsic_shuffler::shuffle::<Block, Self>(&runtime_api, &at,previous_block_extrinsics, seed);
+
+
+							// let extrinsics_hash = BlakeTwo256::hash(&body.clone().encode());
```
Comment (gleb-urvanov) [inline client/service/src/client/client.rs]: remove dead code

File: client/shuffler/Cargo.toml [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+sp-block-builder = "2.0.0"
+sp-inherents = "2.0.0"
+sp-std = { default-features = false, version = '2.0.0' }
+sc-client-api = "2.0.0"
+codec = { package = "parity-scale-codec", version = "1.3.4", features = ["derive"] }
+extrinsic-info-runtime-api = { path='../../primitives/extrinsic-info-runtime-api', version='2.0.0' }
+rand = "0.7.3"
+
+[dev-dependencies]
+#substrate-test-runtime-client = { path = "../../test-utils/runtime/client" }
```
Comment (gleb-urvanov) [inline client/shuffler/Cargo.toml:33]: dead dependency

File: client/shuffler/README.md [hunk 7]
```diff
@@ -0,0 +1,9 @@
+Substrate block builder
+
+This crate provides the [`BlockBuilder`] utility and the corresponding runtime api
```
Comment (gleb-urvanov) [inline client/shuffler/README.md:3]: This readme is unrelated

File: client/shuffler/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+use sp_api::{ApiRef, Encode, HashT, ProvideRuntimeApi};
+use sp_core::crypto::Ss58Codec;
+use sp_core::H256;
+use sp_runtime::generic::BlockId;
+use sp_runtime::traits::{BlakeTwo256, Block as BlockT};
+use sp_runtime::AccountId32;
+use sp_std::collections::btree_map::BTreeMap;
+use sp_std::collections::vec_deque::VecDeque;
+use sp_std::vec::Vec;
+// , HashMap};
```
Comment (gleb-urvanov) [inline client/shuffler/src/lib.rs:12]: dead code

File: client/shuffler/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+use rand::prelude::SliceRandom;
+use rand::rngs::StdRng;
+use rand::SeedableRng;
+
+pub fn shuffle<'a, Block, Api>(
```
Comment (gleb-urvanov) [inline client/shuffler/src/lib.rs:17]: please add /// dovumentation for pub function

File: client/transaction-pool/Cargo.toml [hunk 10]
```diff
@@ -21,23 +21,24 @@ intervalier = "0.4.0"
...
+prometheus-endpoint = { package = "substrate-prometheus-endpoint", version = "0.8.0"}
+sc-client-api = { version = "2.0.0"}
+sc-transaction-graph = { version = "2.0.0"}
+sp-api = { version = "2.0.0"}
+sp-core = { version = "2.0.0"}
+sp-runtime = { version = "2.0.0"}
+sp-tracing = { version = "2.0.0"}
+sp-transaction-pool = { version = "2.0.0"}
+sp-blockchain = { version = "2.0.0"}
+sp-utils = { version = "2.0.0"}
```
Comment (gleb-urvanov) [inline client/transaction-pool/Cargo.toml:33]: same as previously- why use substrate crates while having these locally?
If none of them are modified by us, we should remove them.

File: pallets/xyk/src/lib.rs [hunk 11]
```diff
@@ -29,10 +29,9 @@ pub trait Trait: frame_system::Trait {
 
 const PALLET_ID: ModuleId = ModuleId(*b"79b14c96");
 // 1/100 %
-const TREASURY_PERCENTAGE: u128 = 5; 
-const BUYANDBURN_PERCENTAGE: u128 = 5; 
-const SWAPFEE_PERCENTAGE: u128 = 30; 
-const MANGATA_ID: u128 = 0; 
+const TREASURY_PERCENTAGE: u128 = 5;
+const BUYANDBURN_PERCENTAGE: u128 = 5;
+const MANGATA_ID: u128 = 0;
```
Comment (gleb-urvanov) [inline pallets/xyk/src/lib.rs:34]: Please revert alterations of xyk pallet, since it is not a part of the task

File: primitives/extrinsic-info-runtime-api/src/runtime_api.rs [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+}
+
+sp_api::decl_runtime_apis! {
+    /// The `TaggedTransactionQueue` api trait for interfering with the transaction queue.
+    pub trait ExtrinsicInfoRuntimeApi {
+        /// Validate the transaction.
+        // #[changed_in(2)]
+        // fn get_info(tx: <Block as BlockT>::Extrinsic) -> TransactionValidity;
+
+        /// Validate the transaction.
```
Comment (gleb-urvanov) [inline primitives/extrinsic-info-runtime-api/src/runtime_api.rs:21]: function documentation is unrelated

**Review by gleb-urvanov — CHANGES_REQUESTED** [review]
I'm happy with the proposed solution, and it lays a good way forward for trello#101
Most of the changes requested are cosmetic.

File: client/basic-authorship/Cargo.toml (hunk 1 above)
Comment (mateuszaaa) [reply client/basic-authorship/Cargo.toml:29]: basic authorship was taken from substrate repository where all of the pallets are avaialble locally - all of above are availavle in mangata repository

File: client/transaction-pool/Cargo.toml (hunk 10 above)
Comment (mateuszaaa) [reply client/transaction-pool/Cargo.toml:33]: `Cargo.toml` was taken from substrate repo  where all dependencies are aprt of repository above are not available in mangata repo

**Review by gleb-urvanov — APPROVED** [review]
